modules/library/metadata/src/main/java/org/geotools/factory/FactoryRegistry.java

```diff
@@ -152,15 +158,52 @@ public FactoryRegistry(final Class<?>[] categories) {
      * @param categories The categories.
      */
     public FactoryRegistry(final Collection<Class<?>> categories) {
-        super(categories.iterator());
-        for (final Iterator<Class<?>> it=getCategories(); it.hasNext();) {
-            if (needScanForPlugins == null) {
-                needScanForPlugins = new HashSet<Class<?>>();
-            }
-            needScanForPlugins.add(it.next());
+        for (Class<?> category : categories) {
+            needScanForPlugins.add(category);
+            registry.registerCategory(category);
         }
     }
 
+    /*
+     * TODO: discuss how much API can be broken
+     *
+     * Breakage:
+     *
+     *  - not all methods of `ServiceRegistry` API were implemented
+     *
+     * Possible breakage:
+     *
+     *  - in general nomenclature, replace "ServiceProvider" with "Factory"
+     *  - RegisterableService interface can no longer be used because it expects a `ServiceProvider`
+     *  - replace `Filter` with `Predicate<? super T>`
+     *
+     */
+
+    public Stream<Class<?>> streamCategories() {
+        return registry.streamCategories();
+    }
+
+    // TODO: ensure uniform null handling here and in CategoryRegistry
+
+    // TODO: document
+    public <T> T getServiceProviderByClass(Class<T> providerClass) {
+        if (providerClass == null) {
+            // TODO: do something fancy like in getServiceProvider(Class, Filter, Hints, Hints.Key) ?
+            throw new IllegalArgumentException("The providerClass must not be null.");
+        }
+        return registry.getInstanceOfType(providerClass).orElse(null);
+    }
+
+    // TODO: document
+    public <T> Iterator<T> getServiceProviders(final Class<T> category, final boolean useOrdering) {
```

Comment:
Javadoc the useOrdering parameter with its workflow particularly. Ditto for the filter version below also.